### PR TITLE
feat(mysql): add MySQL database integration

### LINKED ISSUE #530: [FEATURE] Add MySQL integration for database RCA
### Problem statement

MySQL support was part of the original scope in #322  but was deferred  to get early feedback on the PostgreSQL implementation (#515 ). 

This issue tracks the remaining MySQL half.

### Proposed solution

Mirror the PostgreSQL implementation from #515 for MySQL, including  end-to-end tests of a successful investigation flow.

### PR DESCRIPTION
**Pull request by devankitjuneja**
Fixes #530

 

#### Describe the changes you have made in this PR -

Adds full MySQL database integration, mirroring the PostgreSQL integration pattern. Teams with MySQL instances can now use OpenSRE for automated RCA — diagnosing slow queries, checking server health, inspecting replication lag, and analyzing table statistics.

**New files:**
- `app/integrations/mysql.py` — Core integration: `MySQLConfig`, credential resolver (`resolve_mysql_config` reads from store → env → defaults so credentials never appear in LLM tool signatures), 5 diagnostic functions (server status, current processes, replication status, slow queries via `performance_schema`, table stats via `information_schema`)
- `app/tools/MySQL{ServerStatus,CurrentProcesses,ReplicationStatus,SlowQueries,TableStats}Tool/` — 5 `@tool`-decorated wrappers
- `docs/mysql.mdx` + `docs/postgresql.mdx` — Documentation pages for both integrations (registered in `docs/docs.json` nav)

**Modified files:**
- `app/cli/constants.py` — Added `mysql`, `postgresql`, `discord` to `VERIFY_SERVICES` and `SETUP_SERVICES` (these were missing, causing `opensre integrations verify mysql` to reject the argument)
- `app/integrations/cli.py` — `_setup_mysql()` handler + entry in `_HANDLERS`
- `app/integrations/models.py` — `MySQLIntegrationConfig` class + `mysql` field in `EffectiveIntegrations`
- `app/integrations/verify.py` — `_verify_mysql()`, `"mysql"` in `SUPPORTED_VERIFY_SERVICES`
- `app/nodes/resolve_integrations/node.py` — MySQL classification, env loading, service key mapping
- `app/nodes/plan_actions/detect_sources.py` — MySQL source detection from `mysql_database`/`mysql_table` annotations
- `app/types/evidence.py` — `"mysql"` added to `EvidenceSource` Literal
- `tests/tools/conftest.py` — MySQL entry in `mock_agent_state()` fixture

**Tests:** 1973 passing (21 unit tests, 16 e2e tests, 5 tool contract tests)

**Live verification:** Tested against Docker MySQL 8.0 (`docker-compose.testing.yml` `mysql-test` service). `opensre integrations verify mysql` returns `passed — Connected to MySQL 8.0.42`. Investigation pipeline correctly detects `mysql` source from alert annotations.

### Screenshots of the UI changes (If any) -
N/A — backend integration only.

---

## Code Understanding and AI Usage

**Did you use AI assistance (ChatGPT, Claude, Copilot, etc.) to write any part of this code?**
- [ ] No, I wrote all the code myself
- [x] Yes, I used AI assistance (continue below)

**If you used AI assistance:**
- [x] I have reviewed every single line of the AI-generated code
- [x] I can explain the purpose and logic of each function/component I added
- [x] I have tested edge cases and understand how the code handles them
- [x] I have modified the AI output to follow this project's coding standards and conventions

**Explain your implementation approach:**

The MySQL integration follows the exact same two-layer pattern established by the PostgreSQL integration:

1. **Integration module** (`app/integrations/mysql.py`) — handles all MySQL-specific SQL and connection logic. Uses `pymysql` with `DictCursor`. SSL is handled via `ssl.create_default_context()` mapped from the three `ssl_mode` strings (`preferred`/`required`/`disabled`). Replication detection tries `SHOW REPLICA STATUS` first (MySQL 8.0.22+) then falls back to `SHOW SLAVE STATUS`. Slow queries use `performance_schema.events_statements_summary_by_digest` with timer values in picoseconds divided by 1e9 for millisecond precision.

2. **Tool wrappers** — thin `@tool`-decorated functions whose signatures only expose `host`, `database`, `port`. Credentials are resolved server-side by `resolve_mysql_config()` (store → env → defaults), keeping them out of LLM context.

The implementation reuses the same config resolution, store, and verification patterns already proven with PostgreSQL, MongoDB, and MariaDB — no novel abstractions introduced.

---

## Checklist before requesting a review
- [x] I have added proper PR title and linked to the issue
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [x] If it is a core feature, I have added thorough tests
- [x] My code follows the project's style guidelines and conventions

---

Note: Please check **Allow edits from maintainers** if you would like us to assist in the PR.

### CODE REVIEW

File: app/integrations/mysql.py [hunk 1]
```diff
@@ -0,0 +1,575 @@
...
+    ok: bool
+    detail: str
+
+
+def build_mysql_config(raw: dict[str, Any] | None) -> MySQLConfig:
+    """Build a normalized MySQL config object from env/store data."""
+    return MySQLConfig.model_validate(raw or {})
+
+
+def mysql_config_from_env() -> MySQLConfig | None:
```
Comment (Devesh36) [inline app/integrations/mysql.py:80]: mysql_config_from_env and the env loading in node.py both call build_mysql_config which goes through MySQLConfig — but MySQLConfig has a _normalize_password validator that calls .strip() on the password. Stripping whitespace from passwords silently changes credentials if a user has a password with leading/trailing spaces, which is valid in MySQL. This is the same issue present in the PostgreSQL integration and should be fixed across both

Comment (devankitjuneja) [reply app/integrations/mysql.py:80]: Thanks @Devesh36  for pointing this out. Working on its fix.
Also, can you tell me your process of surfacing these bugs early?

**Comment by Devesh36** [comment]
if this worked locally add screenshot for this . Also @devankitjuneja if possible in this pr add docs for this  opensre.com/docs here or go for it in next pr

**Comment by devankitjuneja** [comment]
> if this worked locally add screenshot for this . Also @devankitjuneja if possible in this pr add docs for this opensre.com/docs here or go for it in next pr

docs included in the PR. Will add screenshots shortly

**Comment by Devesh36** [comment]
> > if this worked locally add screenshot for this . Also @devankitjuneja if possible in this pr add docs for this opensre.com/docs here or go for it in next pr
> 
> docs included in the PR. Will add screenshots shortly

yup do it , i'll  merge it asap

**Comment by devankitjuneja** [comment]
image

**Comment by devankitjuneja** [comment]
> > > if this worked locally add screenshot for this . Also @devankitjuneja if possible in this pr add docs for this opensre.com/docs here or go for it in next pr
> > 
> > 
> > docs included in the PR. Will add screenshots shortly
> 
> yup do it , i'll merge it asap

@Devesh36 Done. Please check.
